### ISSUE TITLE
nash panic when It cannot infer NASHPATH, no default value assumed

### ISSUE THREAD

**Issue by ryukinix** [NeowayLabs/nash#260]
# Description

I think we can think a little here about default values, even when there is no $HOME ou there is.

An example was made on building the initiramfs embedded [u-root] busybox system with `nash` just for exploration if nash can be the shell of [u-root]. I fixed that problem defined hard-coded the NASHPATH as `/.nash`, first I tried `/` but happened a really weird bug about init something (I need replicate that issue). 

So then, we need some default value for the worse case that nash cannot infer. In my Linux nash works out of the box, but u-root I need tweaking.  We need think a little about it. 

[u-root]: https://github.com/u-root/u-root

# Testcases

## Out of the box at u-root

![image](https://user-images.githubusercontent.com/7642878/36574082-b43e6006-1822-11e8-982b-8a7bbf0d6425.png)

## After some tweaking

![image](https://user-images.githubusercontent.com/7642878/36574134-edd6d17c-1822-11e8-92fc-a8b681d8bf9f.png)

# TL;DR

We need a default value for NASHPATH and not panic about that. I suggest $HOME/.nash if HOME exists, otherwise `/.nash`, which would be a embedded system that don't have HOME set, probably.

**Comment by katcipis** [comment]
What if we don't have permission to create the /.nash ? them we panic ? Avoiding panic seems really hard...but perhaps some default that does not depends on any env var can be a good idea (even though the default can also fail).

A tmp dir would allow nash to work while there is no env but you won't be able to persist any third party libraries.

There is also the NASHROOT where the stdbin and stdlib is installed, If I'm not wrong it should also be based on user.Current() and it would also fail. We would have to think about it too.

Now that I was thinking about NASHROOT...perhaps when both are not able to be infered nash just runs without any of them...you wont have libs or stdlibs...but it will run and you can use basic builtin stuff...perhaps it is better than panicking.

**Comment by tiago4orion** [comment]
I strongly agree that nash must works out of the box, no need to tweak any environment and so on. But with the introduction of the stdlib (in this case NASHROOT), the user will not be able to use it but I don't see any problem at all.

What do you guys think of the following?

NASHPATH = &lt;user libraries&gt;
We can use &lt;user home&gt;/nash as default and /nash if user's home isnt found.

NASHROOT= &lt;standard libraries&gt;
No default, assume stdlib is embedded in the interpreter binary.

About embedding stdlib in the binary, we have discussed this briefly in the past and I still think it's the best approach because that way we can ship a standalone nash (+ stdlib) binary.

If someone wants just hack the stdlib, he can extract the code by using something like:

```sh
λ> nash -dump-stdlib /home/i4k/nashroot
# hack stdlib...
λ> setenv NASHROOT = "/home/i4k/nashroot"
λ> nash
λ> # now using user's stdlib
```

**Comment by ryukinix** [comment]
I think @tiago4orion is a good solution. But as @katcipis said, I'm still thinking about panic about permissions error if we fix default value, for instance, by `/nash` or even  `$HOME/nash` (which would be a really weird  error). We need think about it, but for me permissions problems is really something that can panic for me, since changing permissions without user permissions is not good.

But there is something that we can do about that without changing permissions? I can't think useful nothing about it for now, but if we will panic about that (permission error etc), we indeed should show a nice error explaining what happens: since the user understand what is necessary to do. How much NASHPATH is vital to Nash works?

Getting a default-value as described on this thread, as well talked before with the u-root devs, nash would be embedded inside of u-root busybox system  https://github.com/u-root/u-root/issues/648

**Comment by katcipis** [comment]
@tiago4orion we would embed the standard binaries on nash too ? I feel that working standalone is great...but to be honest embedding everything just to make this works seems convoluted to me. It is not just hacking the stdlib...understanding what is going on on your code will be hard too. Errors on the stdlib will report which line code ? Some line code inside the binary ? None ? I would be considerably confused if an interpreted language that I'm using has no library anywhere on the system but still seens to work. Of course this can be corrected by the person reading the docs...but I don't know, though choice x_x

Right now I'm more prone to working standalone, just to dont panic, so you can set up proper directories and finish the install. If you don't want to install de stdlib and stdbin..that is ok too, just don't use them and use the core language. Seems less convoluted and provides a way to not panic and make the computed usable if nash is the only default shell installed.

The nash binary itself could help (with --help) on setting the dirs and downloading the standard stuff. Involves more steps but I think it is more explicit. But I ma be tripping here =P. @vitorarins any thoughts ?

**Comment by katcipis** [comment]
Right now this issue can be closed because nash no longer exits if any error related to NASHPATH and NASHROOT happens, it will just work without stdlib and without history. But we have no facility to help install the stdlib after the nash install for now.

**Comment by ryukinix** [comment]
Great to know! I'll test it later. :heart:
